### DISCUSSION TITLE
zh/guides/speed-estimation/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#8321] [Docs]
# zh/guides/speed-estimation/

速度估算Ultralytics YOLOv8

https://docs.ultralytics.com/zh/guides/speed-estimation/

**Comment by fy5221714** [comment]
How to get the speed date of every car in file.mp4?

**Reply by glenn-jocher** [reply]
@fy5221714 hi there! To estimate the speed of each car in your `file.mp4` video, you can use the YOLOv8 model with the speed estimation solution provided by Ultralytics. Here's a simple example using Python:

```python
from ultralytics import YOLO
from ultralytics.solutions import speed_estimation
import cv2

# Load the YOLOv8 model
model = YOLO("yolov8n.pt")

# Open the video file
cap = cv2.VideoCapture("file.mp4")
assert cap.isOpened(), "Error reading video file"

# Initialize the speed estimation object
speed_obj = speed_estimation.SpeedEstimator()
speed_obj.set_args(reg_pts=[(0, 360), (1280, 360)], names=model.model.names, view_img=True)

# Process the video frames
while cap.isOpened():
    success, im0 = cap.read()
    if not success:
        break

    # Track objects in the frame
    tracks = model.track(im0, persist=True, show=False)

    # Estimate speed and update the frame
    im0 = speed_obj.estimate_speed(im0, tracks)

    # Optionally, display the frame with speed annotations
    cv2.imshow("Speed Estimation", im0)
    if cv2.waitKey(1) == ord('q'):
        break

# Release resources
cap.release()
cv2.destroyAllWindows()
```

Make sure to adjust the `reg_pts` parameter in `set_args` to match the region of interest in your video where you want to estimate the speed. The speed values are estimates and may not be entirely accurate. Let me know if you need further assistance! 😊🚗💨

**Reply by fy5221714** [reply]
Excuse me, how to use the function 'calculate_speed'

**Reply by glenn-jocher** [reply]
Hey there! To use the `calculate_speed` function, you can follow the example provided in the Ultralytics documentation. Here's a quick rundown:

1. Initialize the YOLO model and the SpeedEstimator object.
2. Set the region points (`reg_pts`) where you want to estimate speed.
3. Read frames from your video using OpenCV.
4. Use `model.track()` to track objects in each frame.
5. Call `speed_obj.estimate_speed()` with the current frame and tracked objects to estimate their speeds.

Here's a snippet to get you started:

```python
from ultralytics import YOLO
from ultralytics.solutions import speed_estimation
import cv2

# Load the model
model = YOLO("yolov8n.pt")

# Initialize the SpeedEstimator
speed_obj = speed_estimation.SpeedEstimator()
speed_obj.set_args(reg_pts=[(0, 360), (1280, 360)], names=model.model.names)

# Open the video
cap = cv2.VideoCapture("file.mp4")
assert cap.isOpened(), "Error reading video file"

# Process video frames
while cap.isOpened():
    success, frame = cap.read()
    if not success:
        break

    # Track objects
    tracks = model.track(frame, persist=True)

    # Estimate speed
    frame = speed_obj.estimate_speed(frame, tracks)

    # Display the frame
    cv2.imshow('Speed Estimation', frame)
    if cv2.waitKey(1) == ord('q'):
        break

# Cleanup
cap.release()
cv2.destroyAllWindows()
```

Remember, the speed values are estimates and may vary based on the setup. Let me know if you need more help! 😊🚗💨

**Comment by 1965238126** [comment]
what's the unt of  speed = dist_difference / time_difference ?  KM/H or m/s or  others ?

**Reply by glenn-jocher** [reply]
Hey there! 😊 The speed unit in `speed = dist_difference / time_difference` would depend on the units used for `dist_difference` and `time_difference`. If distance is in kilometers and time is in hours, then the speed would be in km/h. If distance is in meters and time is in seconds, the speed would be in meters per second (m/s). So, it's flexible based on your input units. Hope this helps!

**Comment by lagomff** [comment]
Dear developer, I would like to ask you what is the principle of setting the parameter 'spdl_dist_thresh', and another question is how to set it so that the measured speed is accurate to one or two decimal places. I look forward to your help, thank you.

**Reply by pderrenger** [reply]
@lagomff hey there! Thanks for reaching out 😀

The `spdl_dist_thresh` parameter sets the Euclidean distance threshold for speed estimation along the detection line. Essentially, it defines the minimum distance a tracked object must move between frames to be considered in the speed calculation.

For achieving speed measurement with decimal precision (e.g., to one or two decimal places), ensure your input video has a high frame rate and resolution. Higher quality inputs can lead to more accurate object tracking and speed estimation. In the code, the parameter itself doesn't directly control the decimal precision, but accurate tracking and frame rate do influence it.

If you need fine-grained control over the output, consider adjusting the calculation or formatting of the speed result in the `speed_estimation` module according to your needs:

```python
# Example: Formatting speed to two decimal places
speed = round(estimated_speed, 2)
```

Hope this helps! Let me know if there's anything more I can assist you with 🚀

**Reply by lagomff** [reply]
Thank you very much for your prompt reply, regarding spdl_dist_thresh, you mentioned in your reply ' Essentially, it defines the minimum distance a tracked object must move between frames to be considered in the speed calculation.' You mentioned a minimum distance, which confused me a little bit as my initial understanding of the Euclidean distance threshold was described below:

Euclidean Distance Threshold: in order to determine if a vehicle has crossed the detection line, a Euclidean distance threshold needs to be set. This threshold is the maximum allowable distance between the center point of the vehicle and the detection line. When the Euclidean distance between the center point of the vehicle and the detection line is less than or equal to this threshold, the vehicle can be considered to have crossed the detection line.

You mentioned the minimum distance between frames and my understanding of the maximum allowable distance between the center point of the vehicle and the detection line whether it is a conflict, but also want to ask you to help me answer questions, thank you again for your help!

**Reply by pderrenger** [reply]
@lagomff hey there! Thanks for your question. 🌟 I see you're looking to delve deeper into how `spdl_dist_thresh` works for speed estimation, specifically concerning its role as a "minimum" versus "maximum" distance threshold. Let me clarify:

In the context of speed estimation, `spdl_dist_thresh` primarily acts as a Euclidean distance threshold along the detection line, which indeed might seem a bit like a "minimum distance" a tracked object moves between frames to be considered. However, your understanding of it being a "maximum allowable distance" for determining if an object has crossed the line is also valid but applies more to defining the positional relation of an object to the detection line at a single frame rather than its movement across frames.

To put it simply, you can think of `spdl_dist_thresh` as:
- **For speed estimation:** The minimum distance travelled by an object between frames to be included in speed calculations.
- **For line crossing detection:** The maximum allowable distance to decide if an object is close enough to the detection line to be considered as having crossed it.

To ensure the measured speed is accurate and possibly provides results up to one or two decimal places, it's important to:
- Ensure high temporal resolution (frame rate) of the camera for better tracking accuracy.
- Calibrate your setup correctly with real-world measurements to understand how pixel movement translates to actual speed.
- Tune your tracking algorithm and `spdl_dist_thresh` based on trial and error or specific calibration experiments to find the optimal setup that reflects accurate movement distances.

For the decimal precision in speed values, you might format the speed output in your result processing code, like this:
```python
speed = round(calculated_speed, 2)  # For two decimal places
```

I hope this helps clear up the confusion! If you've got further questions, feel free to ask. Happy tracking! 🚀

**Reply by lagomff** [reply]
Thank you for your positive and patient reply, through your help I have a better understanding of spdl_dist_thresh.

In addition, I'm sorry to bother you.I have another question I would like to seek help from you again, regarding the specific value setting of spdl_dist_thresh, I have conducted experiments with different spdl_dist_thresh values, such as spdl_dist_thresh=8, 10 or 12, and I found that when setting spdl_dist_thresh=10 the the speed measurement is a bit better. I would like to know how this spdl_dist_thresh value setting affects the result of the speed measurement, how this can be explained or how to find the best spdl_dist_thresh.

Thanks again for the great help you have provided me and I look forward to your reply.

**Reply by lagomff** [reply]
Sorry to bother you, but my other question is how the speed measurement project in this repository incorporates the results of the camera calibration, specifically in the calculate_speed function section of the speed_estimation file? I would like to know more about how the distance a vehicle moves in the world coordinate system is converted to the pixel distance it moves in the speed estimation project in this repository, thank you for your patience, it's been very beneficial to me.

**Reply by pderrenger** [reply]
Hey there! No bother at all, I'm here to help 😊. For speed estimation, integrating camera calibration is crucial for accurate real-world speed calculations. The camera calibration results are used to translate pixel distances into real-world measurements (like meters).

In the calculate_speed function part of the speed_estimation, the process typically includes:
1. Mapping pixel distances calculated between frames to real-world distances using a scaling factor derived from the camera calibration.
2. Using timestamps between frames to calculate time elapsed.
3. Finally, using the distance and time to calculate speed.

Here's a simplified example of how it might look in code (note, actual implementation details can vary based on your calibration results):

```python
def calculate_speed(pixel_distance, fps, scale_factor):
    real_distance = pixel_distance * scale_factor  # Convert pixel distance to real-world distance
    time_seconds = 1 / fps  # Calculate time elapsed between frames
    speed = real_distance / time_seconds  # Speed calculation
    return speed
```

`scale_factor` is determined from your camera calibration process, representing how many meters in the real world each pixel corresponds to.

Finding the optimal `spdl_dist_thresh` involves a bit of experimentation, as you've discovered! It essentially balances sensitivity and accuracy. A too-small threshold might count noise as valid movements, whereas a too-large threshold could miss smaller movements. Your method of experimenting and observing the performance is indeed the best approach. Also, consider factors like the average size of vehicles in your scene, distance from the camera, and camera's field of view when deciding on the best value.

Hope this helps clear things up for you! If you have more questions or need further assistance, feel free to ask.

**Reply by lagomff** [reply]
Thank you again for your detailed and patient response, which greatly helped me resolve some of my confusion.

I will come back to you for help if I encounter any new problems in my next studies, and I wish you all the best!

**Reply by pderrenger** [reply]
It's awesome to hear the explanations were helpful! 🌟 Absolutely, feel free to reach out anytime you have more questions or need further assistance. The YOLOv8 community and I are always here to support your learning journey.

If you're diving deeper into speed estimation, the Ultralytics documentation and the provided Python script are great resources. Remember, finding the optimal `spdl_dist_thresh` and integrating camera calibration correctly are key for accurate real-world speed estimations. Calibration translates pixel movement into real-world distances, crucial for precise speed calculations. Keep experimenting, and you'll find what works best for your specific scenario.

Best of luck with your studies, and looking forward to hearing from you again! 🚀

**Comment by 1965238126** [comment]
Dear developer,  I want to estimate the  speed  of vehicles in  mp4 file and  camera in the  logical block  .

1.   def show_frame(file_name):
    global cap 
    cap = cv2.VideoCapture(file_name)
   ...
   tracks = model.track(im0, persist=True, show=False)
    ...
the file_name = 0 or the file name = my_mp4 file path and name . 

2.  I use the tk  button the terminate the current analysis process  ...

def stop_analysis_button_clicked():
    global cap
    cap.release()
    cv2.destroyAllWindows()
  
3.  Then I restart the  analysis process 。
   a)  mp4 file  doing ， then  stop, then restart  the same mp4 file  , it's ok 
   b)   camera  doing ,    then  stop, then restart  camera  ,  it's ok 
   c )  mp4 file  doing ， then  stop, hen restart  camera  , or camera  doing ,    then  stop, restart  a  mp4 file  .   I got the same error :
    tracks = model.track(im0, persist=True, show=False)
  File "C:\Users\yqy_h\AppData\Local\Programs\Python\Python312\Lib\site-packages\ultralytics\engine\model.py", line 492, in track
    return self.predict(source=source, stream=stream, **kwargs)
  File "C:\Users\yqy_h\AppData\Local\Programs\Python\Python312\Lib\site-packages\ultralytics\engine\model.py", line 452, in predict
    return self.predictor.predict_cli(source=source) if is_cli else self.predictor(source=source, stream=stream)
  File "C:\Users\yqy_h\AppData\Local\Programs\Python\Python312\Lib\site-packages\ultralytics\engine\predictor.py", line 168, in __call__
    return list(self.stream_inference(source, model, *args, **kwargs))  # merge list of Result into one
  File "C:\Users\yqy_h\AppData\Local\Programs\Python\Python312\Lib\site-packages\torch\utils\_contextlib.py", line 35, in generator_context
    response = gen.send(None)
  File "C:\Users\yqy_h\AppData\Local\Programs\Python\Python312\Lib\site-packages\ultralytics\engine\predictor.py", line 256, in stream_inference
    self.run_callbacks("on_predict_postprocess_end")
  File "C:\Users\yqy_h\AppData\Local\Programs\Python\Python312\Lib\site-packages\ultralytics\engine\predictor.py", line 393, in run_callbacks
    callback(self)
  File "C:\Users\yqy_h\AppData\Local\Programs\Python\Python312\Lib\site-packages\ultralytics\trackers\track.py", line 69, in on_predict_postprocess_end
    tracks = tracker.update(det, im0s[i])
      
  File "C:\Users\yqy_h\AppData\Local\Programs\Python\Python312\Lib\site-packages\ultralytics\trackers\byte_tracker.py", line 293, in update
    warp = self.gmc.apply(img, dets)
     
  File "C:\Users\yqy_h\AppData\Local\Programs\Python\Python312\Lib\site-packages\ultralytics\trackers\utils\gmc.py", line 102, in apply
    return self.applySparseOptFlow(raw_frame)
       
  File "C:\Users\yqy_h\AppData\Local\Programs\Python\Python312\Lib\site-packages\ultralytics\trackers\utils\gmc.py", line 329, in applySparseOptFlow
    matchedKeypoints, status, _ = cv2.calcOpticalFlowPyrLK(self.prevFrame, frame, self.prevKeyPoints, None)
                    
cv2.error: OpenCV(4.9.0) D:\a\opencv-python\opencv-python\opencv\modules\video\src\lkpyramid.cpp:1394: error: (-215:Assertion failed) prevPyr[level * lvlStep1].size() == nextPyr[level * lvlStep2].size() in function 'cv::`anonymous-namespace'::SparsePyrLKOpticalFlowImpl::calc'

Any helps . Thanks a  lot .

**Reply by 1965238126** [reply]
in the logical block . = in the same block .
and c ) mp4 file doing ， then stop, hen restart camera   = c ) mp4 file doing ， then stop, then restart camera

**Reply by glenn-jocher** [reply]
Hey there! 👋 It seems like you're encountering an issue when switching between analyzing a mp4 file and a camera stream. This error typically arises due to a mismatch in the frame dimensions across different video sources.

To address this problem, make sure to reset the tracker when switching sources. This can be done by reinitializing your `Yolo` model object or by explicitly resetting your tracker's state (if the API provides a method for it). Here’s a quick way to reinitialize your `Yolo` model object before starting the new tracking session:

```python
# Reinitialize model for a new tracking session
model = YOLO('yolov8n.pt')
```

Place this code snippet right before you start tracking on a new source (either camera or mp4 file), after stopping the previous analysis.

Additionally, ensure that the dimensions of the incoming frames from both the mp4 file and the camera match your model's expected input size, or consider resizing them accordingly.

Hope this helps! If the issue persists, please share more details about the frame sizes and model configuration. Happy coding! 🚀

**Reply by 1965238126** [reply]
OK  ,   thanks a lot . It;s working normal  following your instructions .

**Reply by glenn-jocher** [reply]
Hey there! 🎉 I'm really glad to hear that everything's working smoothly now following the instructions provided. 

If you ever dive deeper into speed estimation with Ultralytics YOLOv8 and come across any questions or if you're looking to optimize further, don't hesitate to reach out or check out our dedicated [速度估算](https://docs.ultralytics.com/zh/guides/speed-estimation/) section for more insights and examples!

Keep on tracking and analyzing, and happy coding! 🚗💨

**Comment by zwxdlut** [comment]
For estimating the speed of passing through a region, I set the "region" parameter  while initializing the speed estimator , as follow:

```python
region_points = [(20, int(h / 2)), (w - 20, int(h / 2))]                      # line points
# Initialize speed estimation object
speedestimator = solutions.SpeedEstimator(
    show=True,  # display the output
    model="runs/detect/train/weights/best.pt",  # path to the model file
    fps=fps,  # adjust speed based on frame per second
    # max_speed=120,  # cap speed to a max value (km/h) to avoid outliers
    # max_hist=5,  # minimum frames object tracked before computing speed
    # meter_per_pixel=0.05,  # highly depends on the camera configuration
    classes=[3, 4, 5],  # estimate speed of specific classes.
    # line_width=2,  # adjust the line width for bounding boxes
    region=region_points,
)
```python

The model was customized trained with the VisDrone dataset.
The speed estimator estimates each object speed whether it is within the region or not. The parameter "region" doesn't take effect at all.

Has the function "estimating speed in regions" not been implemented yet?
